cmd/start/main.go

```diff
@@ -416,26 +429,125 @@ func checkoutAndAssembleContext(
+}
 
-	// Loop over the directory's files and remove them.
-	for _, f := range dirFiles {
-		filename := filepath.Join(directory, f.Name())
-		err := os.RemoveAll(filename)
+func deleteDirRecursiveWithSkip(root fs.FS, fnSkip FileSkipFunc, fnRemove RemoveFunc) error {
+	return fs.WalkDir(root, ".", func(path string, d fs.DirEntry, err error) error {
 		if err != nil {
-			return fmt.Errorf("could not remove file %s: %w", filename, err)
+			return fmt.Errorf("could not read files in %s: %w", path, err)
+		}
+		if d.IsDir() {
+			if path == "." {
+				return nil
+			}
+			skipFlags := fnSkip(path, d)
+			if skipFlags&leaveAlone != 0 {
+				return fs.SkipDir
+			}
+			if skipFlags&remove != 0 {
+				err = fnRemove(path, true)
+			}
+			if err != nil {
+				return err
+			}
+			if skipFlags&walkChildren != 0 {
+				return nil
+			}
+			return fs.SkipDir
+
 		}
+		if d.Type().IsRegular() {
+			skipFlags := fnSkip(path, d)
+			if skipFlags&remove == 0 {
+				return nil
+			}
+			err = fnRemove(path, false)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+type FileSystemBase struct {
+	filesystem fs.FS
+	base       string
+}
+
+func withBaseFileRemover(base string, removeFunc RemoveFunc) RemoveFunc {
+	return func(path string, isDir bool) error {
+		fspath := filepath.Join(base, path)
+		return removeFunc(fspath, isDir)
 	}
-	return nil
+}
+
+func deleteDirectoryContentsSpareCache(fsb FileSystemBase, fnRemove RemoveFunc) error {
+	// Open the directory and read all its files.
+	cacheDirPath := filepath.Join(".", odsCacheDirName)
+	return deleteDirRecursiveWithSkip(fsb.filesystem, func(path string, d fs.DirEntry) SkipFileRemovalFlags {
+		if path == cacheDirPath {
+			return leaveAlone
+		}
+		return remove
+	}, withBaseFileRemover(fsb.base, fnRemove))
+}
+
+func cleanCache(fsb FileSystemBase, fnRemove RemoveFunc) error {
+	_, err := fsb.filesystem.Open(odsCacheDirName)
+	if err != nil && os.IsNotExist(err) {
+		return nil
+	}
+
+	fsCache, err := fs.Sub(fsb.filesystem, odsCacheDirName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	cacheDependenciesPath := filepath.Join(".", odsCacheDependenciesDirName)
+	// To avoid spare files inside the cache which are not supported delete
+	// all other areas of the cache
+	return deleteDirRecursiveWithSkip(fsCache, func(path string, d fs.DirEntry) SkipFileRemovalFlags {
+		if strings.HasPrefix(path, cacheDependenciesPath) {
+			// Dependencies must be inside a folder specific to a technology
+			// such as for npm or go.
+			// Clean all files which are not directories
+			listOfPath := strings.Split(path, string(os.PathSeparator)) // https://stackoverflow.com/a/33619038
+			if path == cacheDependenciesPath {
+				return walkChildren
+			}
+			if len(listOfPath) == 2 {
+				if d.IsDir() {
+					return leaveAlone
+				} else {
+					return remove
+				}
+			}
+			return remove
+		}
+		return remove // delete everything else
+	}, withBaseFileRemover(filepath.Join(fsb.base, odsCacheDirName), fnRemove))
```

Comment:
I wonder if it would be helpful to move all of your additions to a separate file like `cache.go`, and add some comments there how it works in general to provide a high-level overview?

Reply:
Moved code in these files `cache.go`, `cache_test.go` and also `directory.go`. I was contemplating to move directory into an internal package, but the current code exits on first error and should perhaps have a mode to log or collect such errors instead and continue. Are the comments reflecting what you were looking for? I was aiming to link types of the standard lib in the comments  (reading https://github.com/golang/go/discussions/48305). Let me know if there is a better way.